packages/main/src/popup-utils/PopoverRegistry.js

```diff
@@ -0,0 +1,111 @@
+import { isClickInRect } from "./PopupUtils.js";
+import { getOpenedPopups, addOpenedPopup, removeOpenedPopup } from "./OpenedPopupsRegistry.js";
+
+let updateInterval = null;
+const intervalTimeout = 300;
+const openedRegistry = [];
+
+const repositionPopovers = event => {
+	openedRegistry.forEach(popover => {
+		popover.reposition();
+	});
+};
+
+const attachGlobalScrollHandler = () => {
+	document.body.addEventListener("scroll", repositionPopovers, true);
+};
+
+const detachGlobalScrollHandler = () => {
+	document.body.removeEventListener("scroll", repositionPopovers, true);
+};
+
+const runUpdateInterval = () => {
+	updateInterval = setInterval(() => {
+		repositionPopovers();
+	}, intervalTimeout);
+};
+
+const stopUpdateInterval = () => {
+	clearInterval(updateInterval);
+};
+
+const attachGlobalClickHandler = () => {
+	document.addEventListener("mousedown", clickHandler);
+};
+
+const detachGlobalClickHandler = () => {
+	document.removeEventListener("mousedown", clickHandler);
+};
+
+const clickHandler = event => {
+	const openedPopovers = openedRegistry;
+	const openedPopups = getOpenedPopups();
+
+	if (openedPopups.length === 0 || !(openedPopups[openedPopups.length - 1].openBy)) {
+		return;
+	}
+
+	// loop all open popovers
+	for (let i = (openedPopovers.length - 1); i !== -1; i--) {
+		const popover = openedPopovers[i];
+
+		// if popover is modal, opener is clicked or there is one more popover above, skip closing
+		if (popover.modal || popover.isOpenerClicked(event)) {
+			return;
+		}
+
+		if (isClickInRect(event, popover.getBoundingClientRect())) {
+			break;
+		}
+
+		popover.close();
+	}
+};
+
+const attachScrollHandler = popover => {
+	popover && popover.shadowRoot.addEventListener("scroll", repositionPopovers, true);
+};
+
+const detachScrollHandler = popover => {
+	popover && popover.shadowRoot.removeEventListener("scroll", repositionPopovers);
+};
+
+const addOpenedPopover = instance => {
+	addOpenedPopup(instance);
+	openedRegistry.push(instance);
+
+	attachScrollHandler(instance);
```

Comment:
just to clarify why we need to attach scroll handler per instance + global scroll handler

Reply:
global scroll handler does not bubble outside the shadowroot => its not called. Therefore we need a scroll handler to each instance which should reposition all opened popovers on the page (this is because we need to handle the usa case of having nested scrollable popovers)